common/dpcpp/slab_hash.hpp

```diff
@@ -23,49 +24,118 @@ template <size_t A, size_t B, size_t P> struct DefaultHasher {
 };
 
 template <typename T> struct SlabNode {
-  SlabNode() = default;
   SlabNode(T el) {
     for (int i = 0; i < SLAB_SIZE; i++) {
       data[i] = el;
     }
   }
 
   T data[SLAB_SIZE];
-  sycl::global_ptr<SlabNode<T>> next = nullptr;
+  sycl::device_ptr<SlabNode<T>> next = nullptr;
 };
 
 template <typename T> struct SlabList {
   SlabList() = default;
 
-  sycl::global_ptr<SlabNode<T>> root;
+  sycl::device_ptr<SlabNode<T>> root;
+};
+
+template <typename T> struct HeapMaster {
+  HeapMaster(sycl::queue &q) : _q(q) {
+    _heap = sycl::malloc_device<SlabNode<T>>(CLUSTER_SIZE, q);
+    _head = _heap;
+    sycl::device_ptr<uint32_t> tmp_lock = sycl::malloc_device<uint32_t>(1, q);
+
+    q.single_task([=]() { *tmp_lock = 0; });
+
+    _lock = tmp_lock;
+  }
+
+  ~HeapMaster() { sycl::free(_heap, _q); }
+
+  sycl::device_ptr<SlabNode<T>> malloc_node() {
+    sycl::device_ptr<SlabNode<T>> ret;
+    while (sycl::atomic<uint32_t,
+                        sycl::access::address_space::global_device_space>(_lock)
+               .fetch_or(1)) {
+    }
+    ret = _head;
+    _head++;
+    sycl::atomic<uint32_t, sycl::access::address_space::global_device_space>(
+        _lock)
+        .fetch_and(0);
+    return ret;
+  }
+
+  sycl::device_ptr<uint32_t> _lock = nullptr;
+  sycl::device_ptr<SlabNode<T>> _heap;
+  sycl::device_ptr<SlabNode<T>> _head;
+  sycl::queue &_q;
+};
+template <typename T> struct AllocAdapter {
```

Comment:
I wanted all the data to be created from the outside (so that, for example, a specific iterator is distributed only to certain threads), and the table is created inside
do you propose to change the design to create a table from the outside, and access it from the kernel via a buffer?

Reply:
Buffer usage seem orthogonal to this. "wanted all the data to be created from the outside" -- got it, LGTM